src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2254,79 +2267,84 @@ void InterpreterSelectQuery::addPrewhereAliasActions()
     }
 }
 
-void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum processing_stage, QueryPlan & query_plan)
+/// Based on the query analysis, check if optimizing the count trivial count to use totalRows is possible
+std::optional<UInt64> InterpreterSelectQuery::getTrivialCount(UInt64 max_parallel_replicas)
 {
-    auto & query = getSelectQuery();
     const Settings & settings = context->getSettingsRef();
-
-    /// Optimization for trivial query like SELECT count() FROM table.
     bool optimize_trivial_count =
         syntax_analyzer_result->optimize_trivial_count
-        && (settings.max_parallel_replicas <= 1)
+        && (max_parallel_replicas <= 1)
         && !settings.allow_experimental_query_deduplication
         && !settings.empty_result_for_aggregation_by_empty_set
         && storage
         && storage->getName() != "MaterializedMySQL"
         && !storage->hasLightweightDeletedMask()
         && query_info.filter_asts.empty()
-        && processing_stage == QueryProcessingStage::FetchColumns
```

Comment:
Will queries which read from Distributed tables break or return possibly wrong result?

Reply:
AFAICS no for 2 reasons:
* The checks is kept in the original place (`InterpreterSelectQuery::executeFetchColumns`) before calling `getTrivialCount` during execution. If `processing_stage != QueryProcessingStage::FetchColumns` then it won't use trivial count, as before.

* StorageDistributed can't use trivial count, which means that `getTrivialCount` during analysis will return an empty value, so it won't change any setting.

I might be missing other scenarios though.